### PR TITLE
refactor(Password): Create new password and confirm password components

### LINKED ISSUE #831: refactor(Password): Extract common functionality/template into new component
Create a new component that contains the new password and confirm new password fields. Then we could use this new component in all the places we change a password.

### PR DESCRIPTION
**Pull request by geoffreykwan**
## Changes

Created new password and confirm new password components that is now used by all the places where a new password is requested.

## Test

Make sure the new password components work properly in these locations
- Register student
- Register teacher
- Forgot student password
- Forgot teacher password
- User edit password
- Teacher change student password in Classroom Monitor Manage Students view
- Unlink Google account

Closes #831

### CODE REVIEW

File: src/app/password/new-password-and-confirm/new-password-and-confirm.component.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Component, Input, OnInit } from '@angular/core';
+import { FormGroup } from '@angular/forms';
+
+@Component({
+  selector: 'new-password-and-confirm',
+  templateUrl: './new-password-and-confirm.component.html',
+  styleUrls: ['./new-password-and-confirm.component.scss']
+})
+export class NewPasswordAndConfirmComponent implements OnInit {
+  @Input() confirmPasswordLabel: string = 'Confirm New Password';
```
Comment (hirokiterashima) [inline src/app/password/new-password-and-confirm/new-password-and-confirm.component.ts]: localize?

File: src/app/password/new-password-and-confirm/new-password-and-confirm.component.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  @Input() formGroup: FormGroup;
+  @Input() passwordLabel: string = 'New Password';
```
Comment (hirokiterashima) [inline src/app/password/new-password-and-confirm/new-password-and-confirm.component.ts]: localize?

File: src/app/register/register-student-form/register-student-form.component.html [hunk 3]
```diff
@@ -84,37 +84,11 @@ <h2 class="standalone__title accent" i18n>Create Student Account</h2>
...
-                   formControlName="confirmPassword"
-                   required />
-            <mat-error *ngIf="passwordsFormGroup.controls['confirmPassword'].hasError('required')" i18n>Confirm Password required</mat-error>
-            <mat-error *ngIf="passwordsFormGroup.controls['confirmPassword'].hasError('passwordDoesNotMatch')" i18n>Password does not match</mat-error>
-          </mat-form-field>
-        </p>
+        <new-password-and-confirm
+            [formGroup]="passwordsFormGroup"
+            [passwordLabel]="'Password'"
+            [confirmPasswordLabel]="'Confirm Password'">
```
Comment (hirokiterashima) [inline src/app/register/register-student-form/register-student-form.component.html]: localize?

File: src/app/register/register-teacher-form/register-teacher-form.component.html [hunk 4]
```diff
@@ -113,37 +113,11 @@ <h2 class="standalone__title accent" i18n>Create Teacher Account</h2>
...
-                   formControlName="confirmPassword"
-                   required />
-            <mat-error *ngIf="passwordsFormGroup.controls['confirmPassword'].hasError('required')" i18n>Confirm Password required</mat-error>
-            <mat-error *ngIf="passwordsFormGroup.controls['confirmPassword'].hasError('passwordDoesNotMatch')" i18n>Password does not match</mat-error>
-          </mat-form-field>
-        </p>
+        <new-password-and-confirm
+            [formGroup]="passwordsFormGroup"
+            [passwordLabel]="'Password'"
+            [confirmPasswordLabel]="'Confirm Password'">
```
Comment (hirokiterashima) [inline src/app/register/register-teacher-form/register-teacher-form.component.html]: localize?

File: src/app/password/confirm-new-password/confirm-new-password.component.html [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<p [formGroup]="formGroup">
+  <mat-form-field appearance="fill" fxFlex>
+    <mat-label i18n>{{ label }}</mat-label>
```
Comment (hirokiterashima) [inline src/app/password/confirm-new-password/confirm-new-password.component.html]: What does this do? Looks like there's nothing to internationalize.

**Review by hirokiterashima — COMMENTED** [review]
Functionality seems to work. I had some thoughts while reading the code, added below:

1. Do we need to break up the NewPasswordAndConfirmComponent into NewPasswordComponent and ConfirmNewPasswordComponent? Would there be a case where we use one and not the other? I think the code would be easier to follow if they were all in one component. I'm wondering if we're over-engineering this.

2. For simplicity, should the newPasswordLabel and confirmNewPasswordLabel not be configurable, and fix them to something like "New Password" and "Confirm New Password"? So in this case, the only input to NewPasswordAndConfirmComponent would be the formGroup. This would make the code a bit less flexible, but remove the complexity of passing in the labels as inputs and remove the need to ensure that they are localized everywhere we call them. It looks like the only place this flexibility may be useful is when the teacher changes the student's password, but we can fix that by adding a "New Student Password" label above it.

**Comment by geoffreykwan** [comment]
1. Yea I was thinking about combining them. Right now we always use both at the same time. I'll work on combining them.
2. There are 3 different sets of labels we use

New Password/Confirm New Password (when changing an existing password)
Password/Confirm Password (when creating an account)
New Student Password/Confirm New Student Password (when a teacher changes a student password)

I think it's better if these labels are descriptive and integrated with the input.

File: src/app/password/new-password-and-confirm/new-password-and-confirm.component.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import { FormControl, FormGroup, ValidationErrors, Validators } from '@angular/forms';
+
+@Component({
+  selector: 'new-password-and-confirm',
+  templateUrl: './new-password-and-confirm.component.html',
+  styleUrls: ['./new-password-and-confirm.component.scss']
+})
+export class NewPasswordAndConfirmComponent implements OnInit {
+  static readonly CONFIRM_NEW_PASSWORD_FORM_CONTROL_NAME: string = 'confirmNewPassword';
+  static readonly NEW_PASSWORD_FORM_CONTROL_NAME: string = 'newPassword';
```
Comment (hirokiterashima) [inline src/app/password/new-password-and-confirm/new-password-and-confirm.component.ts:11]: Is there any way to reduce these variable name lengths? They cause line wrapping

File: src/app/password/new-password-and-confirm/new-password-and-confirm.component.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  PASSWORD_MIN_LENGTH: number = 8;
+  PASSWORD_PATTERN: string = '^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])[a-zA-Z0-9]+$';
+
+  confirmNewPasswordFormControl: FormControl;
+  confirmNewPasswordFormControlName: string =
+    NewPasswordAndConfirmComponent.CONFIRM_NEW_PASSWORD_FORM_CONTROL_NAME;
```
Comment (hirokiterashima) [inline src/app/password/new-password-and-confirm/new-password-and-confirm.component.ts:17]: private?

**Review by hirokiterashima — APPROVED** [review]
LGTM. Would be good to refactor the code in the future so that the components don't need to get the form control names.

**Review by hirokiterashima — APPROVED** [review]

**Pull request merged**